### PR TITLE
enable the fallback to cpu for adaptive_avg_pool2d and max_pool2d

### LINKED ISSUE #79: adaptive_avg_pool2d cannot fallback to cpu
`adaptive_avg_pool2d` cannot fallback to cpu since it dose not have "try catch"

### PR DESCRIPTION
**Pull request by chunyuan-w**
fix #79

### CODE REVIEW

**Comment by chunyuan-w** [comment]
@EikanWang @pinzhenx

File: torch_ipex/csrc/cpu/CustomOPs.h [hunk 1]
```diff
@@ -134,11 +174,25 @@ class NewApaptiveAvgPoolingOp : public torch::autograd::Function<NewApaptiveAvgP
...
+      try {
+        if (torch_ipex::check_auto_dnnl() && input.device().type() == c10::DeviceType::DPCPP) {
+          grad_input = torch_ipex::cpu::AtenIpexCPUDev::dil_adaptive_avg_pool2d_backward(grad_output.is_contiguous() ? grad_output : grad_output.contiguous(), input.is_contiguous() ? input : input.contiguous());
+          return {grad_input, at::Tensor()};
+        }
+      } catch(std::exception& e) {
+          #if defined(_DEBUG)
+          TORCH_WARN(e.what());
+          #endif
+        }
```
Comment (EikanWang) [inline torch_ipex/csrc/cpu/CustomOPs.h:186]: Redundant whitespace

File: torch_ipex/csrc/cpu/CustomOPs.h [hunk 2]
```diff
@@ -134,11 +174,25 @@ class NewApaptiveAvgPoolingOp : public torch::autograd::Function<NewApaptiveAvgP
...
+        if (input.device().type() == c10::DeviceType::DPCPP){
```
Comment (EikanWang) [inline torch_ipex/csrc/cpu/CustomOPs.h:187]: Seems like incorrect indent style

**Comment by EikanWang** [comment]
LGTM. Please refine the code style.

**Comment by EikanWang** [comment]
@pinzhenx r u okay with this fix?

**Comment by pinzhenx** [comment]
LGTM

File: tests/cpu/test_lazy_reorder.py [hunk 3]
```diff
@@ -641,6 +641,52 @@ def test_adaptive_avg_pool2d_backward(self):
...
+        adaptive_avg_pool2d = torch.nn.AdaptiveAvgPool2d(6)
+
+        y_cpu = adaptive_avg_pool2d(x_cpu)
+        y_dpcpp = adaptive_avg_pool2d(x_dpcpp)
+
+        self.assertEqual(
+            y_cpu,
+            y_dpcpp)
+
+        self.assertEqual("cpu", y_cpu.device.type)
```
Comment (pinzhenx) [inline tests/cpu/test_lazy_reorder.py:664]: this check could be removed

File: tests/cpu/test_lazy_reorder.py [hunk 4]
```diff
@@ -641,6 +641,52 @@ def test_adaptive_avg_pool2d_backward(self):
...
+        self.assertEqual(device, y_dpcpp.device.type)
+
+    def test_adaptive_avg_pool2d_backward_not_divisible(self):
+        ipex.enable_auto_dnnl()
```
Comment (pinzhenx) [inline tests/cpu/test_lazy_reorder.py:668]: is it possible to merge the forward and backward ut into a single one?

File: tests/cpu/test_lazy_reorder.py [hunk 5]
```diff
@@ -641,6 +641,52 @@ def test_adaptive_avg_pool2d_backward(self):
...
+        # test the fallback to cpu when the input size is not divisible by the output size
+        adaptive_avg_pool2d = torch.nn.AdaptiveAvgPool2d(6)
+
+        y_cpu = adaptive_avg_pool2d(x_cpu).sum()
+        y_dpcpp = adaptive_avg_pool2d(x_dpcpp).sum()
+        y_cpu.backward()
+        y_dpcpp.backward()
+        self.assertEqual(x_cpu.grad, x_dpcpp.grad)
+
+        self.assertEqual("cpu", x_cpu.grad.device.type)
```
Comment (pinzhenx) [inline tests/cpu/test_lazy_reorder.py:685]: ditto

File: torch_ipex/csrc/cpu/CustomOPs.h [hunk 6]
```diff
@@ -97,9 +110,26 @@ class NewMaxPoolingOp : public torch::autograd::Function<NewMaxPoolingOp> {
...
+      
+      try {
+        if (torch_ipex::check_auto_dnnl() && input.device().type() == c10::DeviceType::DPCPP) {
+          grad_input = torch_ipex::cpu::AtenIpexCPUDev::dil_max_pooling_backward(
+              grad_output.is_contiguous() ? grad_output : grad_output.contiguous(), indices.is_contiguous() ? indices : indices.contiguous(), input.is_contiguous() ? input : input.contiguous(), kernel_size, stride, padding, dilation, ceil_mode);
+          return {grad_input, at::Tensor(), at::Tensor(), at::Tensor(), at::Tensor(), at::Tensor()};
+        }
+      } catch(std::exception& e) {
+#if defined(_DEBUG)
+    TORCH_WARN(e.what());
```
Comment (pinzhenx) [inline torch_ipex/csrc/cpu/CustomOPs.h:122]: nit: 4 more spaces needed here

File: torch_ipex/csrc/cpu/CustomOPs.h [hunk 7]
```diff
@@ -116,13 +146,23 @@ class NewApaptiveAvgPoolingOp : public torch::autograd::Function<NewApaptiveAvgP
...
-        at::Tensor output;
-        if (torch_ipex::check_auto_dnnl() && input.device().type() == c10::DeviceType::DPCPP) {
-          output = torch_ipex::cpu::AtenIpexCPUDev::dil_adaptive_avg_pool2d(input.is_contiguous() ? input : input.contiguous(), output_size);
+        try{
+          if (torch_ipex::check_auto_dnnl() && input.device().type() == c10::DeviceType::DPCPP) {
+            return torch_ipex::cpu::AtenIpexCPUDev::dil_adaptive_avg_pool2d(input.is_contiguous() ? input : input.contiguous(), output_size);
+          } 
+        } catch(std::exception& e) {
+#if defined(_DEBUG)
+    TORCH_WARN(e.what());
```
Comment (pinzhenx) [inline torch_ipex/csrc/cpu/CustomOPs.h:155]: nit: 4 spaces here. indent according to the catch block

File: torch_ipex/csrc/cpu/CustomOPs.h [hunk 8]
```diff
@@ -134,8 +174,22 @@ class NewApaptiveAvgPoolingOp : public torch::autograd::Function<NewApaptiveAvgP
...
-      if (torch_ipex::check_auto_dnnl() && input.device().type() == c10::DeviceType::DPCPP) {
-        grad_input = torch_ipex::cpu::AtenIpexCPUDev::dil_adaptive_avg_pool2d_backward(grad_output.is_contiguous() ? grad_output : grad_output.contiguous(), input.is_contiguous() ? input : input.contiguous());
+      try {
+        if (torch_ipex::check_auto_dnnl() && input.device().type() == c10::DeviceType::DPCPP) {
+          grad_input = torch_ipex::cpu::AtenIpexCPUDev::dil_adaptive_avg_pool2d_backward(grad_output.is_contiguous() ? grad_output : grad_output.contiguous(), input.is_contiguous() ? input : input.contiguous());
+          return {grad_input, at::Tensor()};
+        }
+      } catch(std::exception& e) {
+#if defined(_DEBUG)
+    TORCH_WARN(e.what());
```
Comment (pinzhenx) [inline torch_ipex/csrc/cpu/CustomOPs.h:184]: ditto

File: tests/cpu/test_lazy_reorder.py (hunk 4 above)
Comment (chunyuan-w) [reply tests/cpu/test_lazy_reorder.py:668]: Other tests in test_lazy_reorder.py separates the backward from the forward. Should we align with other ut in this file? https://github.com/intel/intel-extension-for-pytorch/blob/1d432c2414889e89231c94c7c80d80930c1792b9/tests/cpu/test_lazy_reorder.py#L711

Comment (pinzhenx) [reply tests/cpu/test_lazy_reorder.py:668]: okay, that's just fine.
This code was copied from `test_mkldnn.py` where we've written lots of copy-pasted code there.

**Comment by pinzhenx** [comment]
LGTM. Thanks!

**Pull request merged**